### PR TITLE
Extract post processor logic out of document builder

### PR DESCRIPTION
**Pull request by hellosnow**
@ansyral @vwxyzh  @vicancy @qinezh @superyyrrzz @chenkennt

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs [hunk 1]
```diff
@@ -30,15 +28,15 @@ public class DocumentBuilder : IDisposable
...
         private readonly CompositionHost _container;
         private readonly BuildInfo _currentBuildInfo =
             new BuildInfo
             {
                 BuildStartTime = DateTime.UtcNow,
                 DocfxVersion = typeof(DocumentBuilder).Assembly.GetCustomAttribute<AssemblyFileVersionAttribute>()?.Version
             };
-        private BuildInfo _lastBuildInfo;
+        private readonly BuildInfo _lastBuildInfo;
+        private readonly HandlePostProcessors _handlePostProcessors;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: using noun, e.g. postProcessorsHandler?

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    using Microsoft.DocAsCode.Exceptions;
+    using Microsoft.DocAsCode.Plugins;
+
+    public class HandlePostProcessors : IDisposable
+    {
+        private readonly List<PostProcessor> _postProcessors;
+
+        public HandlePostProcessors(CompositionHost container, ImmutableArray<string> postProcessorNames)
+        {
+            _postProcessors = GetPostProcessor(container, postProcessorNames);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: null check

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                }
+            }
+        }
+
+        public void PostProcess(Manifest manifest, string outputFolder)
+        {
+            // post process
+            foreach (var postProcessor in _postProcessors)
+            {
+                using (new LoggerPhaseScope($"Process in post processor {postProcessor.ContractName}", false))
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: what's the concern not using LoggerPhaseScope("", true)?

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            return processorList;
+        }
+
+        private static void AddBuildInPostProcessor(List<PostProcessor> processorList)
+        {
+            processorList.Add(
+                new PostProcessor
+                {
+                    ContractName = "html",
+                    Processor = new HtmlPostProcessor
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: also use MEF for HtmlPostProcessor?

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs (hunk 3 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: if true, PerformanceScope loglevel is Diagnostic

---
In reply to: [93849068](https://github.com/dotnet/docfx/pull/1055#discussion_r93849068) [](ancestors = 93849068)

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Composition.Hosting;
+
+    using Microsoft.DocAsCode.Build.Common;
+    using Microsoft.DocAsCode.Common;
+    using Microsoft.DocAsCode.Exceptions;
+    using Microsoft.DocAsCode.Plugins;
+
+    public class HandlePostProcessors : IDisposable
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: >public [](start = 4, length = 6)

internal?

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs (hunk 3 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: should we update PerformanceScope loglevel to Verbose?

---
In reply to: [93849816](https://github.com/dotnet/docfx/pull/1055#discussion_r93849816) [](ancestors = 93849816,93849068)

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs (hunk 4 above)
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: no, this is an internal post processor.

---
In reply to: [93849196](https://github.com/dotnet/docfx/pull/1055#discussion_r93849196) [](ancestors = 93849196)

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs (hunk 3 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: I think it's better to add a overload method to pass in loglevel

---
In reply to: [93849871](https://github.com/dotnet/docfx/pull/1055#discussion_r93849871) [](ancestors = 93849871,93849816,93849068)

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs (hunk 1 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: how about PostpcessorsBuilder...

---
In reply to: [93848200](https://github.com/dotnet/docfx/pull/1055#discussion_r93848200) [](ancestors = 93848200)

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs (hunk 3 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: I will add in separate review.

---
In reply to: [93850170](https://github.com/dotnet/docfx/pull/1055#discussion_r93850170) [](ancestors = 93850170,93849871,93849816,93849068)

File: src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs (hunk 4 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: is it more natural for HtmlPostProcessor to decide it's ContractName, and it's handlers?

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/HandlePostProcessors.cs]: @ansyral I think the built in post processors has sequence and need to insert before all plugins, so we not use MEF here?

---
In reply to: [93850900](https://github.com/dotnet/docfx/pull/1055#discussion_r93850900) [](ancestors = 93850900)

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs (hunk 1 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: prefer handler finally

---
In reply to: [93850674](https://github.com/dotnet/docfx/pull/1055#discussion_r93850674) [](ancestors = 93850674,93848200)

**Comment by qinezh** [comment]
:shipit:

**Pull request merged**